presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeValidator.java

```diff
@@ -217,5 +220,23 @@ private void verifyTypeSignature(VariableReferenceExpression variable, TypeSigna
                 checkArgument(variable.getType().getTypeSignature().equals(actual), "type of variable '%s' is expected to be %s, but the actual type is %s", variable.getName(), variable.getType(), actual);
             }
         }
+
+        private void verifyIntermediateAggregationTypes(Map<VariableReferenceExpression, Aggregation> aggregations)
+        {
+            for (Map.Entry<VariableReferenceExpression, Aggregation> entry : aggregations.entrySet()) {
+                VariableReferenceExpression variable = entry.getKey();
+                Aggregation aggregation = entry.getValue();
+
+                int argumentSize = aggregation.getArguments().size();
+                checkArgument(argumentSize == 1,
+                        "Number of arguments for intermediate aggregation is expected to be 1, got %n", argumentSize);
+
+                TypeSignature expectedTypeSig = variable.getType().getTypeSignature();
+                TypeSignature actualTypeSig = aggregation.getCall().getType().getTypeSignature();
+                checkArgument(
+                        expectedTypeSig.equals(actualTypeSig),
+                        "Return type for intermediate aggregation must be the same as the type of its single argument: expected '%s', got '%s'", expectedTypeSig, actualTypeSig);
```

Comment:
what you are actually checking here is that the return type of the aggregation matches the declared type of the variableReferenceExpression for the aggregation.

Reply:
Good catch!

Reply:
I wonder if we should compare both in the validator?